enhancements/microshift/microshift-router-configuration.md

```diff
@@ -0,0 +1,529 @@
+MicroShift's default ingress router comes as part of the product and is always
+deployed. As part of the start procedure, MicroShift will configure the router
+with fixed parameters, such as being always enabled, and tied to ports 80 and
+443. None of these are configurable for the user, and they also tie to how the
+router is exposed in the cluster.
+
+In order to allow more flexibility, the following configuration parameters shall
+be added:
+- Enable/Disable. Some use cases may be egress only, meaning they do not need
+  an ingress controller. In this case the router pod should not be scheduled
+  to save resources. There should be no rules associated to firewalld or
+  iptables either for an improved security posture.
+- Listening ports. Allow configuring ports other than 80 and 443 improves
+  flexibility.
+- Which IPs the router listens on. Some use cases may require the router to be
+  reachable only from certain networks.
+
+### User Stories
+As a MicroShift user, I want to enable the default router with a configuration
+option.
+
+As a MicroShift user, I want to disable the default router and have all the
+manifests automatically removed.
+
+As a MicroShift user, I want to be able to configure the ports on which the
+default router is listening.
+
+As a MicroShift user, I want to be able to choose the IPs where the router is
+listening.
+
+As a MicroShift user, I want to be able to allow traffic to the router from
+specific IP addresses.
+
+### Goals
+* Allow users to enable the router.
+* Allow users to disable the router.
+* Allow users to reach the router in the configured ports.
+* Allow users to configure in which IPs the router listens.
+* Allow users to allow traffic from specific IP addresses.
+* Allow users to deny traffic to the router.
+* Internal access from applications to the router must remain unchanged when
+  the router is enabled.
+
+### Non-Goals
+N/A
+
+## Proposal
+Each configuration option will get its own section with specifics and
+details.
+As a common baseline, a new top level section will be added to the
+configuration:
+```yaml
+apiServer:
+   ...
+# new from here.
+ingress:
+  status: <Managed|Removed> # Defaults to Managed.
+  ports:
+    http: <int> # Defaults to 80.
+    https: <int> # Defaults to 443.
+  expose: # Defaults to IPs in the host. Details below.
+    - <NIC name|IP address>
+    - ...
+```
+
+### Enable/disable the router
+The following configuration is proposed:
+```yaml
+ingress:
+  status: <Managed|Removed> # Defaults to Managed.
+```
+
+With this option MicroShift will decide whether it should create the router
+upon starting. This includes not just the pod, but all the associated resources
+that come along with it: namespace, services, configmaps, etc.
+
+When setting the option to `Removed`, the next MicroShift restart will
+delete all the default router related resources. Note that any existing routes
+will stop being served and it will require an application deploying an
```

Comment:
This includes existing component routes such as console and auth.  I'm not sure what other options MicroShift offers to replace console and auth, but if the default components need to continue working, it can't delete all the default router related resources.

It will also get rid of some built-in DNS configuration and management and future Gateway API functionality.

Reply:
MicroShift does not have console/auth capabilities. The only APIs from openshift are routes and SCCs. One of the requirements for the feature is to disable the router so that it does not take any resources, we need to remove it to fulfill it (this removes the `openshift-ingress` namespace plus cluster roles and cluster role bindings).